### PR TITLE
V1.3: Edit leave application for approval

### LINKED ISSUE #44: As an elevated user, I can access off day requests
so that I can approve/reject them.

### PR DESCRIPTION
**Pull request by Hafizuddin-github**
**V1.3:**

**Approve/Reject Leave Command added**
- syntax: `approve` or `reject` INDEX
- Example: `approve` 1
- Approve/Reject the selected leave application from the current leave list
- Only higher priority level users can approve lower priority level users
 
**Bug fixed**
- Bug fixed for delete leave command that requires correct priority level.
- Bug fixed for add leave command that unable the user to add leave application with past dates.

Resolves #44

### CODE REVIEW

File: src/main/java/seedu/address/model/ModelManager.java [hunk 1]
```diff
@@ -146,14 +146,14 @@ public void sortEmployee(String field, String order) throws NoEmployeeException
         indicateAddressBookChanged();
     }
```
Comment (Woonhian) [inline src/main/java/seedu/address/model/ModelManager.java:148]: Can remove extra space here

File: src/main/java/seedu/address/model/leave/Leave.java [hunk 2]
```diff
@@ -4,40 +4,48 @@
...
      * Every field must be present and not null.
      */
-    public Leave(Nric employeeId, Date date, Approval approval) {
+    public Leave(EmployeeId employeeId, Date date, Approval approval, PriorityLevel priorityLevel) {
         requireAllNonNull(employeeId, date, approval);
         this.employeeId = employeeId;
         this.date = date;
         this.approval = approval;
+        this.priorityLevel = priorityLevel;
+
```
Comment (Woonhian) [inline src/main/java/seedu/address/model/leave/Leave.java]: Extra space here also

File: src/main/java/seedu/address/model/leave/Leave.java [hunk 3]
```diff
@@ -4,40 +4,48 @@
...
     }
 
     public Approval getApproval() {
         return approval;
     }
 
+    public PriorityLevel getPriorityLevel() {
+        return priorityLevel;
+    }
+
```
Comment (Woonhian) [inline src/main/java/seedu/address/model/leave/Leave.java:47]: Extra space

File: src/test/java/seedu/address/testutil/LeaveListBuilder.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
+package seedu.address.testutil;
+
+import seedu.address.model.LeaveList;
+import seedu.address.model.leave.Leave;
+
+//@@author Hafizuddin-NUS
+/**
+ * A utility class to help with building LeaveList objects.
+ * Example usage: <br>
+ *     {@code LeaveList ll = new LeaveListBuilder().withLeave("REQUEST_1", "REQUEST_1").build();}
```
Comment (Woonhian) [inline src/test/java/seedu/address/testutil/LeaveListBuilder.java:10]: Can backspace the extra spaces here

File: src/main/java/seedu/address/logic/commands/EditLeaveCommand.java [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+
+    public static final String MESSAGE_EDIT_LEAVE_SUCCESS = "Approve/Reject Leave: %1$s";
+    public static final String MESSAGE_INVALID_LEAVE_APPROVAL =
+            "Not authorized to approve leave application.";
+
+    private final Index index;
+    private final EditLeaveDescriptor editLeaveDescriptor;
+
+    /**
+     * @param index               of the leave in the filtered leave list to edit
```
Comment (Woonhian) [inline src/main/java/seedu/address/logic/commands/EditLeaveCommand.java]: Can backspace the extra spaces here

File: src/test/java/seedu/address/testutil/LeaveListBuilder.java (hunk 4 above)
Comment (Hafizuddin-github) [reply src/test/java/seedu/address/testutil/LeaveListBuilder.java:10]: Need space there. Same goes to AddressBookBuilder

File: src/main/java/seedu/address/logic/parser/AddressBookParser.java [hunk 6]
```diff
@@ -135,6 +137,13 @@ public Command parseCommand(String userInput) throws ParseException {
         case ListLeaveCommand.COMMAND_WORD:
             return new ListLeaveCommand();
 
+        case EditLeaveCommand.COMMAND_APPROVE:
+            return new ApproveLeaveCommandParser().parse(arguments);
+
+        case EditLeaveCommand.COMMAND_REJECT:
+            return new RejectLeaveCommandParser().parse(arguments);
+
```
Comment (Woonhian) [inline src/main/java/seedu/address/logic/parser/AddressBookParser.java:155]: Extra spaces here

File: src/test/java/seedu/address/testutil/LeaveUtil.java [hunk 7]
```diff
@@ -23,9 +23,24 @@ public static String getAddLeaveCommand(Leave leave) {
...
+        //sb.append(PREFIX_APPROVAL + leave.getApproval().status + " ");
+        return sb.toString();
+    }
+
+    /**
+     * Returns the part of command string for the given {@code EditLeaveDescriptor}'s details.
+     */
+    public static String getEditLeaveDescriptorDetails(EditLeaveDescriptor descriptor) {
+        StringBuilder sb = new StringBuilder();
+        //descriptor.getNric().ifPresent(nric-> sb.append(PREFIX_NRIC).append(nric.nric).append(" "));
```
Comment (Woonhian) [inline src/test/java/seedu/address/testutil/LeaveUtil.java]: Can put some comment here to indicate the test cases will be updated in future

**Review by Woonhian — APPROVED** [review]
PR approved. @jylee-git please review and merge the PR once you are okay with it, thanks!

**Comment by jinyao-lee** [comment]
@Woonhian noted. Will take a while as the changes are huge

**Review by jinyao-lee — APPROVED** [review]
PR approved once CI tests pass

**Comment by Hafizuddin-github** [comment]
@Woonhian @jylee-git Request for approval for changes made.

**Review by Woonhian — APPROVED** [review]
Very good, PR approved!

**Pull request merged**